rust/src/types/node.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2025 Sonic Operations Ltd
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at soniclabs.com/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+use zerocopy::{FromBytes, Immutable, IntoBytes, Unaligned};
+
+use crate::types::{Commitment, NodeId, Value};
+
+/// A value of a leaf node in a (file based) Verkle trie with its index.
+// NOTE: Changing the layout of this struct will break backwards compatibility of the
+// serialization format.
+#[derive(
+    Debug, Clone, Copy, Default, PartialEq, Eq, FromBytes, IntoBytes, Immutable, Unaligned,
+)]
+#[repr(C)]
+pub struct ValueWithIndex {
+    /// The index of the value in the leaf node.
+    pub index: u8,
+    /// The value stored in the leaf node.
+    pub value: Value,
+}
+
+/// A sparsely populated leaf node in a (file based) Verkle trie.
+// NOTE: This type should NOT implement [`Clone`] because there should never be two instances
+// corresponding to the same logical node.
+// NOTE: Changing the layout of this struct will break backwards compatibility of the
+// serialization format.
+#[derive(Debug, PartialEq, Eq, FromBytes, IntoBytes, Immutable)]
+#[cfg_attr(test, derive(Clone))] // For testing purposes only
+#[repr(C)]
+pub struct SparseLeafNode<const N: usize, C: Commitment> {
+    pub commitment: C,
+    pub stem: [u8; 31],
+    pub values: [ValueWithIndex; N],
+}
+
+impl<const N: usize, C> Default for SparseLeafNode<N, C>
+where
+    C: Commitment + Default,
+{
+    fn default() -> Self {
+        let mut values = [ValueWithIndex::default(); N];
+        values.iter_mut().enumerate().for_each(|(i, v)| {
+            v.index = i as u8;
```

Comment:
Is there really a point to doing this? We'll overwrite this in most cases anyway, no?

Reply:
Hm ok, I still find it somewhat strange that in this hypothetical scenario you might end up with a leaf that contains values `[(42, 0x123), (1, 0x0)]` though.